core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java

```diff
@@ -0,0 +1,76 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.scalar;
+
+import io.trino.operator.aggregation.histogram.SingleTypedHistogram;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.MapBlock;
+import io.trino.spi.function.Convention;
+import io.trino.spi.function.Description;
+import io.trino.spi.function.OperatorDependency;
+import io.trino.spi.function.OperatorType;
+import io.trino.spi.function.ScalarFunction;
+import io.trino.spi.function.SqlType;
+import io.trino.spi.function.TypeParameter;
+import io.trino.spi.type.MapType;
+import io.trino.spi.type.Type;
+import io.trino.type.BlockTypeOperators;
+
+import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.BLOCK_POSITION;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.NULLABLE_RETURN;
+
+@Description("Return a map containing the counts of the elements in the array")
+@ScalarFunction(value = "array_histogram")
+public final class ArrayHistogramFunction
+{
+    private static final int EXPECTED_HISTOGRAM_SIZE = 10;
+
+    private ArrayHistogramFunction() {}
+
+    @TypeParameter("T")
+    @SqlType("map(T, bigint)")
+    public static Block arrayHistogram(
+            @TypeParameter("T") Type elementType,
+            @OperatorDependency(
+                    operator = OperatorType.EQUAL,
+                    argumentTypes = {"T", "T"},
+                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
+            BlockTypeOperators.BlockPositionEqual equalOperator,
+            @OperatorDependency(
+                    operator = OperatorType.HASH_CODE,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
+            @TypeParameter("map(T, bigint)") MapType mapType,
+            @SqlType("array(T)") Block arrayBlock)
+    {
+        SingleTypedHistogram histogram = new SingleTypedHistogram(
+                elementType,
+                equalOperator,
+                hashCodeOperator,
+                EXPECTED_HISTOGRAM_SIZE);
+        int positionCount = arrayBlock.getPositionCount();
+        for (int position = 0; position < positionCount; position++) {
+            if (!arrayBlock.isNull(position)) {
+                histogram.add(position, arrayBlock, 1L);
+            }
+        }
+        BlockBuilder blockBuilder = mapType.createBlockBuilder(null, histogram.getPositionCount());
```

Comment:
Good question. I'll explain my thought process and please let me know anywhere I'm misunderstanding.

My thought was that `expectedEntries` here refers to the number of elements which we expect to insert into the `blockBuilder` being created on this line. Within the `serialize` function on the next line (72), we write each value into the `blockBuilder`. I believe the `getPositionCount` function on the histogram returns the number of keys in the histogram. So given this, my thought was that I can improve the performance of the `blockBuilder` since on line 71 I expect that `getPositionCount` entries will be inserted into the `blockBuilder` in the `serialize` call on line 72.

If I specified `expectedEntries` as `1`, then I expect the `blockBuilder` would need to resize multiple times if there were many keys in the histogram.

Any corrections welcome, thanks!

Reply:
Looked into it more closely, I wasn't sure whether the expectedEntries propagated into the key and value builder, but it does. So using `histogram.getPositionCount()` is better.